distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso

```diff
@@ -1059,6 +1059,138 @@ type Table
+         them must be satisfied for the two rows to be correlated (so
+         effectively a conjunction of all conditions is used). Each condition
+         can either be a `Join_Condition` or an `Expression`.
+       - right_prefix: The prefix added to right table column names in case of
+         name conflict. See "Column Renaming" below for more information.
+       - on_problems: Specifies how to handle problems if they occur, reporting
+         them as warnings by default.
+
+         - If a column name cannot be found, a `No_Such_Column` is reported
+           and an empty result is reported.
+         - If a column index is invalid, an `Index_Out_Of_Bounds_Error` is
+           reported and an empty result is reported.
+         - If a join condition correlates columns whose types are not compatible
+           (for example comparing numeric types with text), a
+           `Invalid_Value_Type` is reported.
+         - If an expression does not start with a `left` or `right` qualifier,
+           but refers to a column which exists in both the left and right table,
+           an `Ambiguous_Column_Reference` is reported.
+
+         In any of the above cases, if a problem occurs, the resulting table
+         will have the desired structure, but it will be empty to indicate that
+         the join has failed due to an erroneous join condition.
+
+       ? Column Renaming
+
+         If a column name is unique across the joined tables, it is kept as is.
+         If columns from the two tables have colliding names, a prefix (by
+         default `Right_`) is added to the name of the column from the right
+         table. The left column remains unchanged. It is possible that there will
+         still be name collisions after adding the prefix - for example if both
+         tables have a column named `A` but one of the tables also _already_ has
+         a column `Right_A`. In that case the name collisions are resolved using
+         the regular renaming strategy - by adding subsequent `_1`, `_2` etc.
+         suffixes to the column names - the original columns have precedence in
+         their name being preserved without change and the columns that were
+         being renamed due to inter-table collisions are renamed in the order of
+         their appearance in the right table.
+
+       ? Expression Syntax
+
+         The expressions provided for the join conditions may use a special
+         syntax `[left.A]` and `[right.A]` to refer to columns from each table
+         that have the same names (the column references from before
+         consolidating column names are used). If a column name `[B]` without
+         the prefix is used, it must be unambiguous (i.e. it must be present in
+         exactly one of the tables) - in any other case, an error will be
+         reported.
+    join2 : Table -> Join_Kind -> Expression | Join_Condition | Vector (Expression | Join_Condition) -> Text -> Problem_Behavior -> Table
```

Comment:
I used name `join2` to avoid clashes with the old join - I do not remove it as I don't want to break other stuff + new join prototype will mostly re-use the old one.

Of course finally old one will be removed and the new one will use the name `join`.